File: client/database/question.model.ts
import { Schema, model, models, Model, Types } from 'mongoose'
import { IUser } from './user.model'
import { ITag } from './tag.model'

export interface IQuestion extends Document {
  title: string
  content: string
  tags: Types.ObjectId[] | ITag[]
  views: number
  upvotes: Types.ObjectId[]
  downvotes: Types.ObjectId[]
  author: Types.ObjectId | IUser
  answers: Types.ObjectId[]
  createdAt: Date
}

const QuestionSchema = new Schema({
  title: { type: String, required: true },
  content: { type: String, required: true },
  tags: [{ type: Schema.Types.ObjectId, ref: 'Tag' }],
  views: { type: Number, default: 0 },
  upvotes: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  downvotes: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  author: { type: Schema.Types.ObjectId, ref: 'User' },
  answers: [{ type: Schema.Types.ObjectId, ref: 'Answer' }],
  createdAt: { type: Date, default: Date.now },
})

export const Question: Model<IQuestion> = models.Question || model<IQuestion>('Question', QuestionSchema)

File: api/internal/data/Question.model.go
package data

import "time"

type Question struct {
	Id           int64     `json:"_id"`
	Title        string    `json:"title"`
	Content      string    `json:"content"`
	Tags         []Tag     `json:"tags"`
	Views        int       `json:"views"`
	Author       User      `json:"author"`
	AuthorId     int64     `json:"authorId"`
	AnswersCount int       `json:"answersCount"`
	CreatedAt    time.Time `json:"createdAt"`
	Upvotes      []int64   `json:"upvotes"`
	Downvotes    []int64   `json:"downvotes"`
}

type CreateQuestionParams struct {
	Id       int64
	Title    string
	Content  string
	AuthorId int64
	Tags     []string
}

type QuestionVoteParams struct {
	Id         int64
	UserId     int64
	QuestionId int64
}
